src/analysis/processing/qgsalgorithmrasterrank.cpp

```diff
@@ -0,0 +1,309 @@
+    {
+      QgsRasterProjector *projector = new QgsRasterProjector();
+      projector->setCrs( rLayer->crs(), outputCrs, context.transformContext() );
+      projector->setInput( rLayer->dataProvider() );
+      projector->setPrecision( QgsRasterProjector::Exact );
+      inputInterfaces[rLayer->id()].reset( projector );
+    }
+    else
+    {
+      inputInterfaces[rLayer->id()].reset( rLayer->dataProvider()->clone() );
+    }
+    inputBlocks[rLayer->id()] = std::make_unique<QgsRasterBlock>();
+  }
+
+  std::unique_ptr<QgsRasterIterator> rasterIterator = std::make_unique<QgsRasterIterator>( inputInterfaces[mLayers.at( 0 )->id()].get() );
+  rasterIterator->startRasterRead( 1, cols, rows, outputExtent );
+  int rasterIteratorCount = static_cast<int>( rasterIterator->blockCount() );
+
+  std::vector<std::unique_ptr<QgsRasterBlock>> outputBlocks;
+  for ( int i = 0; i < mRanks.size(); i++ )
+  {
+    outputBlocks.push_back( std::make_unique<QgsRasterBlock>() );
+  }
+
+
+  const double step = rasterIteratorCount > 0 ? 100.0 / rasterIteratorCount : 0;
+  for ( int i = 0; i < rasterIteratorCount; i++ )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+    feedback->setProgress( i * step );
+
+    int iterLeft = 0;
+    int iterTop = 0;
+    int iterCols = 0;
+    int iterRows = 0;
+    QgsRectangle blockExtent;
+    rasterIterator->next( 1, iterCols, iterRows, iterLeft, iterTop, blockExtent );
+
+    for ( const auto &inputInterface : inputInterfaces )
+    {
+      inputBlocks[inputInterface.first].reset( inputInterface.second->block( 1, blockExtent, iterCols, iterRows ) );
+    }
+
+    for ( int i = 0; i < mRanks.size(); i++ )
+    {
+      outputBlocks[i].reset( new QgsRasterBlock( outputDataType, iterCols, iterRows ) );
+      outputBlocks[i]->setNoDataValue( outputNoData );
+    }
+
+    for ( int row = 0; row < iterRows; row++ )
+    {
+      for ( int col = 0; col < iterCols; col++ )
+      {
+        QList<double> values;
+        for ( const auto &inputBlock : inputBlocks )
+        {
+          bool isNoData = false;
+          const double value = inputBlock.second->valueAndNoData( row, col, isNoData );
+          if ( !isNoData )
+          {
+            values << value;
+          }
+          else if ( outputNoDataOverride )
+          {
+            values.clear();
+            break;
+          }
+        }
+        std::sort( values.begin(), values.end() );
+
+        for ( int i = 0; i < mRanks.size(); i++ )
+        {
+          if ( values.size() >= std::abs( mRanks[i] ) )
+          {
+            outputBlocks[i]->setValue( row, col, values.at( mRanks[i] > 0 ? mRanks[i] - 1 : values.size() + mRanks[i] ) );
+          }
+          else
+          {
+            outputBlocks[i]->setValue( row, col, outputNoData );
+          }
+        }
+      }
+    }
+
+    for ( int i = 0; i < mRanks.size(); i++ )
+    {
+      provider->writeBlock( outputBlocks[i].get(), i + 1, iterLeft, iterTop );
```

Comment:
writeBlock can fail:
```suggestion
      if ( !provider->writeBlock( outputBlocks[i].get(), i + 1, iterLeft, iterTop ) )
      {
        throw QgsProcessingException( QObject::tr( "Could not write output raster block" ) );
      }
```
( i don't think any more detailed errors are available for this right now)